### PR TITLE
AI Pathfinder: do not allow to pass through the well-guarded objects

### LINKED ISSUE #7503: The AI creates path through pyramid, attacks it and dies.
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

The AI green player avoids the blue player by creating path through the pyramid and attacks it, but it is obviously too weak and dies.

It has a decent enough army:
image

https://github.com/ihhub/fheroes2/assets/12501091/c141cf27-1eca-430c-80aa-3ac373c2087a

There are several issues here, but the core one is that the AI's evaluation of the pyramid needs improvement.

### Save file

[ai suicide pyramid.zip](https://github.com/ihhub/fheroes2/files/12240464/ai.suicide.pyramid.zip)

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #7503

Alternative approach to #7553.

Weak blue hero now goes around the Graveyard:

https://github.com/ihhub/fheroes2/assets/32623900/7867ff93-bf42-4216-813b-30d84d93c57e

while powerful green hero just goes through:

https://github.com/ihhub/fheroes2/assets/32623900/1e0e1c84-61f6-4e5f-b9fc-869ad0ab3500

#7553 has the advantage that hero still can go through the protected object by declining a battle (and save a few movement points), but this PR is a bit simpler. Also I'm not sure that we are able to decline a battle in all protected objects (although it seems to be so).

### CODE REVIEW

**Comment by ihhub** [comment]
I am marking this PR for 1.0.8 as it is not a game breaking issue.

**Comment by LeHerosInconnu** [comment]
The AI should always take the shortest path (to save movement points) and not engage in combat when taking the shortest path if this is not its original goal (its original goal being to take the shortest path to a chosen destination).
This is also the behavior of Heroes 2's AI.

**Comment by oleg-derevenetz** [comment]
> This is also the behavior of Heroes 2's AI.

In original Heroes 2, AI even passes through impenetrable guard if it wants to take a "shortest path" to a mighty artifact:

https://www.youtube.com/watch?v=IY9CRBt6OOY

:)

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> > This is also the behavior of Heroes 2's AI.
> 
> In original Heroes 2, AI even passes through impenetrable guard if it wants to take a "shortest path" to a mighty artifact:
> 
> https://www.youtube.com/watch?v=IY9CRBt6OOY
> 
> :)

Yes, I've already seen this video.
But it has more to do with an outrageous AI bonus, than with practicing the shortest path. :)

**Comment by oleg-derevenetz** [comment]
> But it has more to do with an outrageous AI bonus, than with practicing the shortest path. :)

OG AI is also very stupid in its intention to always follow the shortest path. Consider:

https://github.com/ihhub/fheroes2/assets/32623900/850e5d86-a889-409b-b859-13556e0d29f5

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz,

> > But it has more to do with an outrageous AI bonus, than with practicing the shortest path. :)
> 
> OG AI is also very stupid in its intention to always follow the shortest path. Consider:
> 
> https://github.com/ihhub/fheroes2/assets/32623900/850e5d86-a889-409b-b859-13556e0d29f5
> 
> BYW it's at least one of the cases where you cannot decline a battle.
> 
> With this PR:
> 
> https://github.com/ihhub/fheroes2/assets/32623900/0891e3ce-17b8-4c6a-836e-829ce8443146

In the original game, the active square is marked as a move (the mouse cursor is an arched horse), even though combat is automatically engaged.

In fheroes2, the active square of a mine is marked directly as a combat for the human player (the mouse cursor is a sword).
Isn't this the case for AI (without the cursor display)?

In original game:

Before elemental mine 02

In fheroes2:

Before elemental mine 01

Hera are the save files and the scenario file:
[Before elemental mine 01.zip](https://github.com/ihhub/fheroes2/files/12346250/Before.elemental.mine.01.zip)

In your example from the original post, rather than avoiding the active graveyard square, the weak AI hero should move to the active graveyard square, not engage in combat, and continue on his way.

**Comment by oleg-derevenetz** [comment]
> Isn't this the case for AI (without the cursor display)?

I do not know exactly how the AI in the original game treats this tile, but it obviously should not have paved its path through it.

File: src/fheroes2/world/world_pathfinding.cpp [hunk 1]
```diff
@@ -203,8 +194,22 @@ namespace
...
 
-        // If none of the special cases apply, check if tile can be moved on
-        return !MP2::isNeedStayFront( objectType );
+        // If we can't step on this tile, then it cannot be passed through
+        if ( MP2::isNeedStayFront( objectType ) ) {
+            return false;
+        }
+
+        // If we can step on this tile, but it is protected by monsters, then it can be passed through if we manage to defeat the monsters
+        if ( MP2::isProtectedObject( objectType ) ) {
```
Comment (idshibanov) [inline src/fheroes2/world/world_pathfinding.cpp:203]: Some objects are fine to go through if you can decline the combat as long it is the shortest path. I understand your approach but not sure this check will work.

I think we need both improved pathfinding and teach AI to decline.

**Comment by oleg-derevenetz** [comment]
Closing in favor of #7553.

**Pull request closed without merging**